File: lexer.py
import sys


# lexer helper methods

keywords = {}
symbols = {
    "{" : "left-brace",
    "}" : "right-brace",
    "(" : "left-paren",
    ")" : "right-paren",
    "[" : "left-bracket",
    "]" : "right-bracket",
    "<" : "less-than",
    ">" : "greater-than",
    "+" : "plus",
    "-" : "minus",
    "*" : "star",
    "/" : "slash",
    "%" : "percent",
    "!" : "not",
    "?" : "question",
    "=" : "equals",
    "." : "dot",
    "," : "comma",
    "&" : "and", 
    "|" : "or",
    ";" : "semicolon",
    ":" : "colon"
}

maybe_multichar = {
    "=", "!", ">", "<", "+", "-", "*", "/", "%"
}

multi = {
    "==" : "equals-equals",
    "!=" : "not-equals",
    ">=" : "greater-equals",
    "<=" : "less-equals",
    "+=" : "plus-equals",
    "-=" : "minus-equals",
    "*=" : "star-equals",
    "/=" : "slash-equals",
    "%=" : "percent-equals"
}


def is_id_char(c):
    """returns true if the passed character is a valid identifier character"""
    return c.isalnum() or (c == "_")


def lex_string(source):
    global keywords
    global symbols
    
    tokens = []
    i = 0

    while(i < len(source)):

        if source[i].isspace():
            while i < len(source) and source[i].isspace():
                i += 1
        else:
            # identifiers
            if source[i].isalpha():
                j = i + 1
                while j < len(source) and is_id_char(source[j]):
                    j += 1

                word = source[i : j]
                
                if word in keywords.keys():
                    tokens.append( ('keyword', word) )
                else:
                    tokens.append( ('identifier', source[i : j]) )
                i = j
            
            elif source[i].isnumeric():
                j = i + 1
                while j < len(source) and source[j].isnumeric():
                    j += 1
                if source[j] == ".":
                    j += 1
                    while j < len(source) and source[j].isnumeric():
                        j += 1
                    tokens.append( ('real-literal', source[i : j]) )
                    i = j
                else:
                    tokens.append( ('int-literal', source[i : j]) )
                    i = j

            elif source[i] in "\'\"":
                quote = source[i]
                j = i + 1
                while source[j] != quote:
                    j += 1
                tokens.append( ('string-literal', source[ i + 1 : j ]) )
                i = j + 1

            elif source[i] in symbols.keys():
                
                j = i + 2
                if j <= len(source) and source[i : j] in multi.keys():
                    symbol = source[i : j]
                    tokens.append( (multi[symbol], symbol) )
                    i = j
                else:
                    symbol = source[i : i+1]
                    tokens.append( (symbols[symbol], symbol) )
                    i += 1
            else:
                raise Exception("Invalid token start: " + source[i])
    
    print(tokens)
                
                 


if __name__ == "__main__":

    opts = [opt for opt in sys.argv[1:] if opt.startswith("-")]
    args = [arg for arg in sys.argv[1:] if not arg.startswith("-")]

    
    if "-c" in opts:
        if len(args) == 0:
            raise SystemExit("Command mode requires at least one argument")
        else:
            for arg in args:
                lex_string(arg)
    
    # otherwise read in a file and lex that
    else:
        pass

File: lexer.cpp
#include <stdio.h>
#include <ctype.h>
#include <vector> 
#include <string>
#include <unordered_set> 



std::unordered_set<std::string> keywords        ( {"var", "if", "then", "else", "for", "while", "do", "in"} );
std::unordered_set<std::string> symbols         ( {"{", "}", "(", ")", "[", "]", "<", ">", "+", "-", "*", "/", "%", "!", "?", "=", ".", ",", "&", "|", ";", ":" } );
std::unordered_set<std::string> maybe_multichar ( {"=", "!", ">", "<", "+", "-", "*", "/", "%"} );
std::unordered_set<std::string> multi           ( { "==", "!=", ">=", "<=", "+=", "-=", "*=", "/=", "%=" } );

bool is_id_char(char c) {
    return isdigit(c) || isalpha(c) || c == '_';
}

std::vector<std::pair<std::string, std::string>> lex_string ( std::string &source ) {

    std::vector<std::pair<std::string, std::string>> tokens;
    int index = 0;

    while ( index < source.size() ) {
        
        if ( isspace(source[index]) ) {
            // could treat newlines special, but don't feel like it yet
            while( index < source.size() && isspace(source[index]) ) {
                ++index;
            }
        }
        else {

            // identifiers and keywords (reserved words)
            if ( isalpha(source[index]) || source[index] == '_' ) {
                int end = index + 1;

                // move ahead until the end of the identifier
                while ( end < source.size() && is_id_char(source[end]) ) {
                    ++end;
                }
                std::string word = source.substr(index, end - index);
                std::pair<std::string, std::string> token;

                // unordered_set.contains() only came out in c++20
                if ( keywords.count(word) > 0 ) {
                    token = std::make_pair( word, "keyword" ); 
                }
                else {
                    token = std::make_pair( word, "identifier" );
                }
                tokens.push_back(token);
                index = end;
            }
            // integers and real numbers (floating point)
            else if ( isdigit(source[index]) ) {
                int end = index + 1;

                while ( end < source.size() && isdigit(source[end]) ) {
                    ++end;
                }
                std::pair<std::string, std::string> token;

                if ( source[end] == '.' ) {
                    ++end;
                    while ( end < source.size() && isdigit(source[end]) ) {
                        ++end;
                    }
                    std::string real = source.substr(index, end - index);
                    token = std::make_pair(real, "real-literal");
                }
                else {
                    std::string integer = source.substr(index, end - index);
                    token = std::make_pair(integer, "int-literal");
                }
                tokens.push_back(token);
                index = end;
            }
            // string literals
            else if ( source[index] == '\"' || source[index] == '\'' ) {
                char quote = source[index];
                int end = index + 1;
                while ( end < source.size() && source[end] != quote ) {
                    ++end;
                }
                std::string string_literal = source.substr(index + 1, end - (index + 1));
                std::pair<std::string, std::string> token = std::make_pair(string_literal, "string-literal");
                tokens.push_back(token);
            }
            // symbols and operators
            else if (symbols.count(std::string(1, source[index])) > 0) {
                int end = index + 2;
                std::pair<std::string, std::string> token;
                if ( end <= source.size() && ( multi.count(source.substr(index, end - index)) > 0) ) {
                    std::string symbol = source.substr(index, end - index);
                    token = std::make_pair(symbol, "symbol");
                    index = end;
                }
                else {
                    token = std::make_pair(std::string(1, source[index]), "symbol");
                    ++index;
                }
                tokens.push_back(token);
            }
            else {
                printf("Invalid start of token %c\n", source[index]);
                break;
            }

        }
    }
    return tokens;
}

void pretty_print( std::vector<std::pair<std::string, std::string>> &tokens) {
    printf("[\n");

    for (int i = 0; i < tokens.size(); ++i) {
        const char* token = tokens[i].first.c_str();
        const char* token_type = tokens[i].second.c_str();
        printf("%s : %s\n", token, token_type);
    }

    printf("]\n");
}


int main( int argc, const char* argv[] ) {
    // for now just take an argument string and tokenize it, worry about file IO later
    if( argc > 1 ) {
        std::string source = argv[1];

        std::vector<std::pair<std::string, std::string>> tokens = lex_string(source);

        pretty_print(tokens);
    }
    return 0;
}